pkg/config/remote/uptane/transactional_store.go

```diff
@@ -0,0 +1,227 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package uptane
+
+import (
+	"sync"
+
+	"github.com/pkg/errors"
+	"go.etcd.io/bbolt"
+)
+
+// dbBucket contains the data of the bucket
+// if the content it's nil it means the file must be deleted, otherwise it contains data
+type dbBucket map[string][]byte
+
+// transactionalStore persists all the target files present in the current director targets.json
+// All writes go to the in-memory structure until the `commit` method is called.
+// All reads first read the in-memory data and fallback to the on disk one.
+// A call to `commit` will flush all changes to the DB in a single transaction.
+type transactionalStore struct {
+	// underlying database where we apply many changes atomically
+	db *bbolt.DB
+
+	// lock to access the cached data and bbolt db
+	lock sync.RWMutex
```

Comment:
It's probably not really all that high-impact, but we have a lot of locks in our arch - at each layer. One in the remote config client itself, one in the list that is used to track clients, one in the uptane client, and now one here. I suspect that this makes some of them somewhat redundant.

I think this is partly because a lot of these locks are embedded within the structures themselves, rather than being at the layer that actually uses the structure. There's no parallelism internal to the structure, and thus it all depends on how the caller uses it as to whether or not there are locking concerns. Since the uptane client owns the transactional store, and also has a lock itself, I suspect we're already locking sufficiently at that level.

Whether or not this is worth refactoring, I'm not sure. 😄